File: b90/ebpf/profiler.go
package ebpf

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"os"
	"time"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/perf"
	"github.com/cilium/ebpf/ringbuf"
)

//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang -cflags "-O2 -g -Wall" bpf profiler.c

type EventType uint32

const (
	EventIORead    EventType = 1
	EventIOWrite   EventType = 2
	EventMemAlloc  EventType = 3
	EventTCPTx     EventType = 4
	EventTCPRx     EventType = 5
	EventLockWait  EventType = 6
)

type BPFEvent struct {
	Timestamp  uint64
	PID        uint32
	TID        uint32
	Comm       [16]byte
	EventType  uint32
	DurationNs uint64
	Bytes      uint64
	Address    uint64
	Retval     int32
	Filename   [256]byte
	StackID    uint64
}

type ProcessMetrics struct {
	PID            uint32
	IOReadBytes    uint64
	IOWriteBytes   uint64
	PageCacheHits  uint64
	TCPTxBytes     uint64
	TCPRxBytes     uint64
}

type BPFProfiler struct {
	objs         bpfObjects
	links        []link.Link
	ringbuf      *ringbuf.Reader
	eventChan    chan BPFEvent
	metricsChan  chan ProcessMetrics
	stopChan     chan struct{}
}

func NewBPFProfiler() (*BPFProfiler, error) {
	objs := bpfObjects{}
	if err := loadBpfObjects(&objs, nil); err != nil {
		return nil, fmt.Errorf("loading objects: %v", err)
	}

	profiler := &BPFProfiler{
		objs:        objs,
		eventChan:   make(chan BPFEvent, 1024),
		metricsChan: make(chan ProcessMetrics, 1024),
		stopChan:    make(chan struct{}),
	}

	if err := profiler.attachKprobes(); err != nil {
		objs.Close()
		return nil, err
	}

	rb, err := ringbuf.NewReader(objs.Events)
	if err != nil {
		objs.Close()
		return nil, fmt.Errorf("creating ringbuf reader: %v", err)
	}
	profiler.ringbuf = rb

	go profiler.pollEvents()
	go profiler.collectMetrics()

	return profiler, nil
}

func (p *BPFProfiler) attachKprobes() error {
	kp, err := link.Kprobe("vfs_read", p.objs.KprobeVfsRead, nil)
	if err != nil {
		return fmt.Errorf("attaching vfs_read: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kretprobe("vfs_read", p.objs.KretprobeVfsRead, nil)
	if err != nil {
		return fmt.Errorf("attaching vfs_read ret: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kprobe("vfs_write", p.objs.KprobeVfsWrite, nil)
	if err != nil {
		return fmt.Errorf("attaching vfs_write: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kretprobe("vfs_write", p.objs.KretprobeVfsWrite, nil)
	if err != nil {
		return fmt.Errorf("attaching vfs_write ret: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kprobe("__kmalloc", p.objs.KprobeKmalloc, nil)
	if err != nil {
		return fmt.Errorf("attaching kmalloc: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kretprobe("__kmalloc", p.objs.KretprobeKmalloc, nil)
	if err != nil {
		return fmt.Errorf("attaching kmalloc ret: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kprobe("tcp_sendmsg", p.objs.KprobeTcpSendmsg, nil)
	if err != nil {
		return fmt.Errorf("attaching tcp_sendmsg: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kprobe("tcp_recvmsg", p.objs.KprobeTcpRecvmsg, nil)
	if err != nil {
		return fmt.Errorf("attaching tcp_recvmsg: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kprobe("mutex_lock", p.objs.KprobeMutexLock, nil)
	if err != nil {
		return fmt.Errorf("attaching mutex_lock: %v", err)
	}
	p.links = append(p.links, kp)

	kp, err = link.Kretprobe("mutex_lock", p.objs.KretprobeMutexLock, nil)
	if err != nil {
		return fmt.Errorf("attaching mutex_lock ret: %v", err)
	}
	p.links = append(p.links, kp)

	return nil
}

func (p *BPFProfiler) pollEvents() {
	for {
		select {
		case <-p.stopChan:
			return
		default:
			record, err := p.ringbuf.Read()
			if err != nil {
				if err == ringbuf.ErrClosed {
					return
				}
				continue
			}

			var event BPFEvent
			if err := binary.Read(bytes.NewBuffer(record.RawSample), binary.LittleEndian, &event); err != nil {
				continue
			}

			p.eventChan <- event
		}
	}
}

func (p *BPFProfiler) collectMetrics() {
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-p.stopChan:
			return
		case <-ticker.C:
			var pid uint32
			iter := p.objs.ProcessIoRead.Iterate()
			for iter.Next(&pid, nil) {
				var readBytes, writeBytes, cacheHits, txBytes, rxBytes uint64
				
				p.objs.ProcessIoRead.Lookup(&pid, &readBytes)
				p.objs.ProcessIoWrite.Lookup(&pid, &writeBytes)
				p.objs.ProcessPageCache.Lookup(&pid, &cacheHits)
				p.objs.TcpTxBytes.Lookup(&pid, &txBytes)
				p.objs.TcpRxBytes.Lookup(&pid, &rxBytes)

				p.metricsChan <- ProcessMetrics{
					PID:           pid,
					IOReadBytes:   readBytes,
					IOWriteBytes:  writeBytes,
					PageCacheHits: cacheHits,
					TCPTxBytes:    txBytes,
					TCPRxBytes:    rxBytes,
				}
			}
		}
	}
}

func (p *BPFProfiler) Events() <-chan BPFEvent {
	return p.eventChan
}

func (p *BPFProfiler) Metrics() <-chan ProcessMetrics {
	return p.metricsChan
}

func (p *BPFProfiler) GetStack(stackID uint32) ([]uint64, error) {
	var stack [128]uint64
	err := p.objs.StackTraces.Lookup(stackID, &stack)
	if err != nil {
		return nil, err
	}
	
	var result []uint64
	for _, addr := range stack {
		if addr == 0 {
			break
		}
		result = append(result, addr)
	}
	return result, nil
}

func (p *BPFProfiler) Close() {
	close(p.stopChan)
	if p.ringbuf != nil {
		p.ringbuf.Close()
	}
	for _, l := range p.links {
		l.Close()
	}
	p.objs.Close()
	close(p.eventChan)
	close(p.metricsChan)
}

File: b90/ebpf/profiler.c
//go:build ignore

#include "vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>

#define TASK_COMM_LEN 16
#define MAX_STACK_DEPTH 128
#define SQL_BUF_SIZE 2048

char LICENSE[] SEC("license") = "Dual BSD/GPL";

struct event {
    u64 ts;
    u32 pid;
    u32 tid;
    char comm[TASK_COMM_LEN];
    u32 event_type;
    u64 duration_ns;
    u64 bytes;
    u64 address;
    int retval;
    char filename[256];
    u64 stack_id;
};

struct {
    __uint(type, BPF_MAP_TYPE_RINGBUF);
    __uint(max_entries, 1 << 24);
} events SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 8192);
    __type(key, u64);
    __type(value, u64);
} io_start_times SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 8192);
    __type(key, u64);
    __type(value, u64);
} mem_alloc_times SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_STACK_TRACE);
    __uint(key_size, sizeof(u32));
    __uint(value_size, MAX_STACK_DEPTH * sizeof(u64));
    __uint(max_entries, 1024);
} stack_traces SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 8192);
    __type(key, u32);
    __type(value, u64);
} process_page_cache SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 8192);
    __type(key, u32);
    __type(value, u64);
} process_io_read SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 8192);
    __type(key, u32);
    __type(value, u64);
} process_io_write SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 8192);
    __type(key, u32);
    __type(value, u64);
} tcp_tx_bytes SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 8192);
    __type(key, u32);
    __type(value, u64);
} tcp_rx_bytes SEC(".maps");

static __always_inline void get_process_comm(char *comm) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    BPF_CORE_READ_STR_INTO(comm, task, comm);
}

SEC("kprobe/vfs_read")
int BPF_KPROBE(kprobe_vfs_read, struct file *file, char *buf, size_t count) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u64 ts = bpf_ktime_get_ns();
    
    bpf_map_update_elem(&io_start_times, &pid_tgid, &ts, BPF_ANY);
    return 0;
}

SEC("kretprobe/vfs_read")
int BPF_KRETPROBE(kretprobe_vfs_read, int ret) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u64 *start_ts = bpf_map_lookup_elem(&io_start_times, &pid_tgid);
    
    if (start_ts && ret > 0) {
        u64 duration = bpf_ktime_get_ns() - *start_ts;
        
        struct event *e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
        if (e) {
            e->ts = bpf_ktime_get_ns();
            e->pid = pid;
            e->tid = (u32)pid_tgid;
            e->event_type = 1;
            e->duration_ns = duration;
            e->bytes = ret;
            e->retval = ret;
            get_process_comm(e->comm);
            e->stack_id = bpf_get_stackid(ctx, &stack_traces, BPF_F_USER_STACK);
            bpf_ringbuf_submit(e, 0);
        }
        
        u64 *bytes = bpf_map_lookup_elem(&process_io_read, &pid);
        u64 new_bytes = bytes ? *bytes + ret : ret;
        bpf_map_update_elem(&process_io_read, &pid, &new_bytes, BPF_ANY);
    }
    
    bpf_map_delete_elem(&io_start_times, &pid_tgid);
    return 0;
}

SEC("kprobe/vfs_write")
int BPF_KPROBE(kprobe_vfs_write, struct file *file, const char *buf, size_t count) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u64 ts = bpf_ktime_get_ns();
    
    bpf_map_update_elem(&io_start_times, &pid_tgid, &ts, BPF_ANY);
    return 0;
}

SEC("kretprobe/vfs_write")
int BPF_KRETPROBE(kretprobe_vfs_write, int ret) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u64 *start_ts = bpf_map_lookup_elem(&io_start_times, &pid_tgid);
    
    if (start_ts && ret > 0) {
        u64 duration = bpf_ktime_get_ns() - *start_ts;
        
        struct event *e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
        if (e) {
            e->ts = bpf_ktime_get_ns();
            e->pid = pid;
            e->tid = (u32)pid_tgid;
            e->event_type = 2;
            e->duration_ns = duration;
            e->bytes = ret;
            e->retval = ret;
            get_process_comm(e->comm);
            bpf_ringbuf_submit(e, 0);
        }
        
        u64 *bytes = bpf_map_lookup_elem(&process_io_write, &pid);
        u64 new_bytes = bytes ? *bytes + ret : ret;
        bpf_map_update_elem(&process_io_write, &pid, &new_bytes, BPF_ANY);
    }
    
    bpf_map_delete_elem(&io_start_times, &pid_tgid);
    return 0;
}

SEC("kprobe/__kmalloc")
int BPF_KPROBE(kprobe_kmalloc, size_t size, gfp_t flags) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u64 ts = bpf_ktime_get_ns();
    
    bpf_map_update_elem(&mem_alloc_times, &pid_tgid, &ts, BPF_ANY);
    return 0;
}

SEC("kretprobe/__kmalloc")
int BPF_KRETPROBE(kretprobe_kmalloc, void *ret) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u64 *start_ts = bpf_map_lookup_elem(&mem_alloc_times, &pid_tgid);
    
    if (start_ts && ret) {
        u64 duration = bpf_ktime_get_ns() - *start_ts;
        
        struct event *e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
        if (e) {
            e->ts = bpf_ktime_get_ns();
            e->pid = pid;
            e->tid = (u32)pid_tgid;
            e->event_type = 3;
            e->duration_ns = duration;
            e->address = (u64)ret;
            get_process_comm(e->comm);
            bpf_ringbuf_submit(e, 0);
        }
    }
    
    bpf_map_delete_elem(&mem_alloc_times, &pid_tgid);
    return 0;
}

SEC("kprobe/tcp_sendmsg")
int BPF_KPROBE(kprobe_tcp_sendmsg, struct sock *sk, struct msghdr *msg, size_t size) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    
    u64 *bytes = bpf_map_lookup_elem(&tcp_tx_bytes, &pid);
    u64 new_bytes = bytes ? *bytes + size : size;
    bpf_map_update_elem(&tcp_tx_bytes, &pid, &new_bytes, BPF_ANY);
    
    struct event *e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
    if (e) {
        e->ts = bpf_ktime_get_ns();
        e->pid = pid;
        e->tid = (u32)pid_tgid;
        e->event_type = 4;
        e->bytes = size;
        get_process_comm(e->comm);
        bpf_ringbuf_submit(e, 0);
    }
    
    return 0;
}

SEC("kprobe/tcp_recvmsg")
int BPF_KPROBE(kprobe_tcp_recvmsg, struct sock *sk, struct msghdr *msg, size_t size, int flags) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    
    u64 *bytes = bpf_map_lookup_elem(&tcp_rx_bytes, &pid);
    u64 new_bytes = bytes ? *bytes + size : size;
    bpf_map_update_elem(&tcp_rx_bytes, &pid, &new_bytes, BPF_ANY);
    
    struct event *e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
    if (e) {
        e->ts = bpf_ktime_get_ns();
        e->pid = pid;
        e->tid = (u32)pid_tgid;
        e->event_type = 5;
        e->bytes = size;
        get_process_comm(e->comm);
        bpf_ringbuf_submit(e, 0);
    }
    
    return 0;
}

SEC("kprobe/mutex_lock")
int BPF_KPROBE(kprobe_mutex_lock, struct mutex *lock) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u64 ts = bpf_ktime_get_ns();
    
    bpf_map_update_elem(&io_start_times, &pid_tgid, &ts, BPF_ANY);
    return 0;
}

SEC("kretprobe/mutex_lock")
int BPF_KRETPROBE(kretprobe_mutex_lock, int ret) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u64 *start_ts = bpf_map_lookup_elem(&io_start_times, &pid_tgid);
    
    if (start_ts) {
        u64 duration = bpf_ktime_get_ns() - *start_ts;
        
        struct event *e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
        if (e) {
            e->ts = bpf_ktime_get_ns();
            e->pid = pid;
            e->tid = (u32)pid_tgid;
            e->event_type = 6;
            e->duration_ns = duration;
            get_process_comm(e->comm);
            bpf_ringbuf_submit(e, 0);
        }
    }
    
    bpf_map_delete_elem(&io_start_times, &pid_tgid);
    return 0;
}

SEC("kprobe/page_cache_ra_unbounded")
int BPF_KPROBE(kprobe_page_cache, struct file *filp, struct address_space *mapping,
                pgoff_t index, unsigned long req_count, int ra_pages) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    
    u64 *count = bpf_map_lookup_elem(&process_page_cache, &pid);
    u64 new_count = count ? *count + 1 : 1;
    bpf_map_update_elem(&process_page_cache, &pid, &new_count, BPF_ANY);
    
    return 0;
}
